### PR TITLE
[AutoParallel][PIR] Support split bw for pir

### PR DESCRIPTION
**Pull request by AndSonder**
### PR Category

Auto Parallel

### PR Types

Performance

### Description

为 PIR 支持 split_bw 策略

测试环境  4卡 A100 
llama2 13b 纯 pp4 vpp2 num_hidden_layers hack 为 8，其他配置详见下面的 PaddleNLP 测试脚本

性能对比：

5bfa66367aa0b4e9853c12d037f0bba3

性能指标都是 interval _tokens_per_second_per_device

pir 下的这个提升是符合预期的

Loss 收敛曲线（对比开启 split_bw 和不开启 split_bw）,由于改变了梯度累计顺序，loss 收敛曲线符合预期

image

测试脚本如下：

```
set -x
unset CUDA_VISIBLE_DEVICES
export FLAGS_enable_pir_api=1
export GLOG_v=0

to_static=1
task_name="llama_13b_pir_perf_tostatic_${to_static}"
log_dir="log/$task_name"
output_dir="output/$task_name"
rm -rf $log_dir
rm -rf $output_dir

# export PARALLEL_CROSS_ENTROPY=true
export FLAGS_call_stack_level=2
export PYTHONPATH=../../../../:$PYTHONPATH
# export PYTHONPATH=/root/paddlejob/workspace/env_run/liyamei/Paddle/build_0926/python/:$PYTHONPATH
# export NCCL_BUFFSIZE=33554432
export FLAGS_force_sync_ops=", , , embedding_grad"
export FLAGS_call_stack_level=3

use_pir=1
max_steps=1000
export FLAGS_enable_pir_api=${use_pir}
export FLAGS_enable_pir_in_executor=1

python -u -m paddle.distributed.launch \
    --gpus=0,1,2,3 \
    --log_dir ${log_dir} \
    ../run_pretrain_auto.py \
    --to_static ${to_static} \
    --enable_auto_parallel true \
    --skip_memory_metrics 0 \
    --output_dir ${output_dir} \
    --model_name_or_path "meta-llama/Llama-2-13b" \
    --tokenizer_name_or_path "meta-llama/Llama-2-13b" \
    --input_dir "../data" \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 4 \
    --tensor_parallel_degree 1 \
    --pipeline_parallel_degree 4 \
    --logging_steps 1 \
    --max_steps $max_steps \
    --save_steps 5000 \
    --eval_steps 1000 \
    --max_seq_length 4096 \
    --learning_rate 3e-05 \
    --min_learning_rate 3e-06 \
    --warmup_steps 30 \
    --weight_decay 0.01 \
    --sequence_parallel 0 \
    --device "gpu" \
    --warmup_ratio 0.01 \
    --dataloader_num_workers 1 \
    --continue_training 0 \
    --do_train true \
    --do_eval false \
    --do_predict false \
    --disable_tqdm true \
    --skip_profile_timer true \
    --save_total_limit 2 \
    --recompute false \
    --recompute_use_reentrant true \
    --distributed_dataloader 0 \
    --recompute_granularity "full" \
    --gradient_accumulation_steps 4 \
    --bf16 1 \
    --fp16_opt_level "O2" \
    --amp_custom_black_list "reduce_sum" "c_softmax_with_cross_entropy" \
    --amp_custom_white_list "lookup_table" "lookup_table_v2" \
    --amp_master_grad true \
    --max_grad_norm 0.0 \
    --use_flash_attention true \
    --fuse_attention_ffn true \
    --fuse_attention_qkv true \
    --use_fused_rope true \
    --use_fused_rms_norm false \
    --fused_linear_param_grad_add true \
    --pipeline_schedule_mode "VPP" \
    --virtual_pp_degree 2 \
    --tensor_parallel_config "enable_mp_async_allreduce" \
    --pipeline_parallel_config "enable_send_recv_overlap enable_split_backward" \
    --data_parallel_config "enable_allreduce_avg_in_gradinent_scale gradient_sync_after_accumulate" \
    --num_hidden_layers 8 \
```

Pcard-76459

### CODE REVIEW

**Review by JZ-LIANG — APPROVED** [review]
LGTM
